File: javascript/lib/gherkin/token_formatter_builder.js
module.exports = function TokenFormatterBuilder() {
  var tokensText = '';

  this.startRule = function(ruleType) {};

  this.endRule = function(ruleType) {};

  this.build = function(token) {
    tokensText += formatToken(token) + '\n';
  };

  this.getResult = function() {
    return tokensText;
  }

  function formatToken(token) {
    if(token.isEof) return 'EOF';

    return "(" +
    token.location.line +
    ":" +
    token.location.column +
    ")" +
    token.matchedType +
    ":" +
    (token.matchedKeyword === undefined ? '' : token.matchedKeyword) +
    "/" +
    (token.matchedText === undefined ? '' : token.matchedText) +
    "/" +
    token.matchedItems.map(function (i) { return i.column + ':' + i.text; }).join(',');
  }

  function format(format) {
    var args = Array.prototype.slice.call(arguments, 1);
    return format.replace(/{(\d+)}/g, function(match, number) {
      return typeof args[number] != 'undefined'
      ? args[number]
      : match
      ;
    });
  };
};

File: elixir/lib/gherkin/token_formatter_builder.ex
defmodule Gherkin.TokenFormatterBuilder do

  def new do
    %{tokens_text: []}
  end

  def reset(builder), do: new

  def start_rule(builder, _), do: builder

  def end_rule(builder, _), do: builder

  def build(builder = %{tokens_text: tokens_text}, token) do
    %{builder | tokens_text: [format_token(token)|tokens_text]}
  end

  def get_result(%{tokens_text: tokens_text}) do
    tokens_text
    |> Enum.reverse
    |> Enum.join("\n")
  end

  defp format_token(%{line: nil}), do: "EOF"
  defp format_token(token) do
    location = "(#{token.location.line}:#{token.location.column})"
    "#{location}#{token.matched_type}:#{token.matched_keyword}/#{token.matched_text}/#{matched_items(token)}"
  end

  defp matched_items(%{matched_items: nil}), do: ""
  defp matched_items(%{matched_items: items}) do
    items
    |> Enum.map(fn (%{column: c, text: t}) -> "#{c}:#{t}" end)
    |> Enum.join(",")
  end

end
